src/Umbraco.Forms.Integrations.Commerce.EMerchantPay/PaymentProviderWorkflow.cs

```diff
@@ -0,0 +1,211 @@
+#else
+        public override WorkflowExecutionStatus Execute(Record record, RecordEventArgs e)
+#endif
+        {
+            if (!CustomerDetailsMappings.TryParseMappings(out var mappings)) return WorkflowExecutionStatus.Failed;
+
+            var mappingBuilder = new MappingBuilder()
+#if NETCOREAPP
+                .SetValues(context.Record, mappings)
+#else
+                .SetValues(record, mappings)
+#endif
+                .Build();
+
+            // step 1. Create or Retrieve Consumer
+            var consumer = new ConsumerDto { Email = mappingBuilder.Email };
+
+            // step 1. Create Consumer
+            var createConsumerTask = Task.Run(async () => await _consumerService.Create(consumer));
+
+            var result = createConsumerTask.Result;
+            if (result.Code == Constants.ErrorCode.ConsumerExists)
+            {
+                // step 1.1. Get Consumer
+                var retrieveConsumerTask = Task.Run(async () => await _consumerService.Retrieve(consumer));
+                consumer = retrieveConsumerTask.Result;
+            }
+
+            // step 2. Create Payment
+            var transactionId = Guid.NewGuid();
+
+            var payment = new PaymentDto
+            {
+                TransactionId = transactionId.ToString(),
+                Usage = Usage,
+                NotificationUrl = "https://www.example.com/notification",
+                ReturnSuccessUrl = "https://www.example.com/success",
+                ReturnFailureUrl = "https://www.example.com/failure",
+                ReturnCancelUrl = "https://www.example.com/cancel",
+                Amount = int.Parse(Amount),
+                Currency = Currency,
+                ConsumerId = consumer.Id,
+                CustomerEmail = consumer.Email,
+                CustomerPhone = mappingBuilder.Phone,
+                BillingAddress = new AddressDto
+                {
+                    FirstName = mappingBuilder.FirstName,
+                    LastName = mappingBuilder.LastName,
+                    Address1 = mappingBuilder.Address,
+                    Address2 = string.Empty,
+                    ZipCode = mappingBuilder.ZipCode,
+                    City = mappingBuilder.City,
+                    State = mappingBuilder.State,
+                    Country = mappingBuilder.Country
+                },
+                BusinessAttribute = new BusinessAttribute { NameOfTheSupplier = Supplier },
+                TransactionTypes = new TransactionTypeDto
+                {
+                    TransactionTypes = new List<TransactionTypeRecordDto>
+                    {
+                        new TransactionTypeRecordDto { TransactionType = "authorize" },
+                        new TransactionTypeRecordDto { TransactionType = "sale" }
+                    }
+                }
+            };
+
+            var createPaymentTask = Task.Run(async () => await _paymentService.Create(payment));
```

Comment:
Remember we decided that...

> we should focus on the hosted payment page, as that seems most useful for our use case of "people without a full e-commerce site, but who want to take payments for things like a one-off event etc."

Looks like you are trying to do the actual payment transaction within the workflow.  I'm not averse to having that too, but I think it's less useful than the option where we use their hosted payment page.

Have a read back over the "Hosted Payment Page" section of the document I shared for how I think this should work - you'll need to investigate the details/ask for documentation.  But I'm expecting we'll should be making a call here to their API and get back a URL to redirect to, passing the details of the price, currency and any personal details that have been collected and mapped.  

And then on that page the user will pay and will be redirect back to one of the three URLs we've provided (success, failed or cancelled).

Finally on the notification URL we'll need to receive a request indicating if a successful payment was made, so we can update the record data.

Reply:
We need to set up or retrieve a customer by email, then create a payment with the mandatory fields, before actually receiving the redirect URL to the hosted payment page.